src/ripple/overlay/Slot.h

```diff
@@ -0,0 +1,635 @@
+        protocol::MessageType type);
+
+    /** Called when a peer is deleted. If the peer is selected to be the
+     * source of messages from the validator then unsquelch handler
+     * is called for all currently squelched peers.
+     * @param id Peer's id
+     */
+    void
+    unsquelch(id_t id);
+
+    /** Check if peers stopped relaying messages
+     * and if slots stopped receiving messages from the validator.
+     */
+    void
+    checkIdle();
+
+    /** Return number of peers in state in state */
+    boost::optional<std::uint16_t>
+    inState(PublicKey const& validator, PeerState state) const
+    {
+        auto const& it = slots_.find(validator);
+        if (it != slots_.end())
+            return it->second.inState(state);
+        return {};
+    }
+
+    /** Return number of peers in state in state */
+    boost::optional<std::uint16_t>
+    notInState(PublicKey const& validator, PeerState state) const
+    {
+        auto const& it = slots_.find(validator);
+        if (it != slots_.end())
+            return it->second.notInState(state);
+        return {};
+    }
+
+    /** Return true if Slot is in State */
+    template <typename Comp = std::equal_to<>>
+    boost::optional<bool>
+    inState(PublicKey const& validator, SlotState state, Comp comp = {}) const
+    {
+        auto const& it = slots_.find(validator);
+        if (it != slots_.end())
+            return comp(it->second.state_, state);
+        return {};
+    }
+
+    /** Get selected peers */
+    std::set<id_t>
+    getSelected(PublicKey const& validator)
+    {
+        auto const& it = slots_.find(validator);
+        if (it != slots_.end())
+            return it->second.getSelected();
+        return {};
+    }
+
+    /** Get peers info. Return map of peer's state, count, and squelch
+     * expiration milliseconds.
+     */
+    std::unordered_map<
+        typename Peer::id_t,
+        std::tuple<PeerState, uint16_t, uint32_t, std::uint32_t>>
+    getPeers(PublicKey const& validator)
+    {
+        auto const& it = slots_.find(validator);
+        if (it != slots_.end())
+            return it->second.getPeers();
+        return {};
+    }
+
+    /** Get Slot's state */
+    boost::optional<SlotState>
+    getState(PublicKey const& validator)
+    {
+        auto const& it = slots_.find(validator);
+        if (it != slots_.end())
+            return it->second.getState();
+        return {};
+    }
+
+private:
+    /** Called when a peer is deleted. If the peer was selected to be the
+     * source of messages from the validator then squelched peers have to be
+     * unsquelched.
+     * @param id Peer's id
+     * @param erase If true then erase the peer
+     */
+    void
+    deletePeer(id_t id, bool erase);
+
+    hash_map<PublicKey, Slot<clock_type>> slots_;
+    SquelchHandler const& handler_;  // squelch/unsquelch handler
+};
+
+template <typename clock_type>
+void
+Slots<clock_type>::checkForSquelch(
```

Comment:
A function called `checkForX` sounds like a pure function that either returns a `bool` or an `X`. Consider re-naming this.